ecs-init/volumes/state_manager.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package volumes
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+)
+
+const (
+	// PluginStatePath is the directory path to the plugin state information
+	PluginStatePath = "/var/lib/ecs/data/"
+	// PluginStateFile contains the state information of the plugin
+	PluginStateFile = "ecs_volume_plugin.json"
+	// PluginStateFileAbsPath is the absolute path of the plugin state file
+	PluginStateFileAbsPath = "/var/lib/ecs/data/ecs_volume_plugin.json"
+)
+
+var lock sync.Mutex
+
+// SaveState saves a representation of 'a' to the file at path
+func SaveState(a interface{}) error {
```

Comment:
I think we should have this method take a strongly typed object instead of relying on the Volume object. The data persistence layer should have a different model. The two benefits we get from this:

1) you don't need to export all of the fields of Volume in the main method
2) we don't need to worry about how changing Volume in future impacts serialization -- they're decoupled

The downside here is that we need to make changes in two places if we want more data to be persisted -- but, this may also be desirable as it is more explicit. IMO, its worth it given our experiences with managing state in agent.